File: 魔改RC4/origin.cpp

#include <iostream>
#include <stdio.h>  
#include <string.h>  

template<class T> inline void
swap(T& i, T& j)
{
    T tmp = i;
    i = j;
    j = tmp;
}
class RC4
{
public:
    void SetKey(char* K, int keylen);
    void Transform(char* output, const char* data, int len);
private:
    unsigned char S[256];
};

//初始化S盒
void RC4::SetKey(char* K, int keylen)
{
    for (int i = 0; i < 256; i++)
    {
        S[i] = i;
    }
    int j = 0;
    for (int i = 0; i < 256; i++)
    {
        j = (j + S[i] + K[i % keylen]) % 256;
        swap(S[i], S[j]);
    }
}
//生成密钥流
void RC4::Transform(char* output, const char* data, int len)
{
    int i = 0, j = 0;
    unsigned char key;
    for (int k = 0; k < len; k++)
    {
        i = (i + 1) % 256;
        j = (j + S[i]) % 256;
        swap(S[i], S[j]);
        key = S[(S[i] + S[j]) % 256];
        output[k] = key ^ data[k];
    }
}

int main()
{
    char data[256] = "wednesday is shit";
    char output[256] = "";
    printf("明文:%s\n", data);
    char K[] = "iamevilcalf";
    int keylen = sizeof(K);
    RC4 rc4encrypt, rc4decrypt;
    rc4encrypt.SetKey(K, keylen);
    rc4decrypt.SetKey(K, keylen);
    rc4encrypt.Transform(output, data, strlen(data));
    printf("密文: %s\n", output);
    rc4decrypt.Transform(output, output, strlen(data));
    printf("解密后明文: %s\n", output);
    return 0;
}

File: 魔改RC4/origin.py
class RC4:
    def __init__(self):
        self.S = []
    
    def set_key(self, K: bytes, keylen: int):
        """ 初始化S盒 """
        self.S = list(range(256))
        j = 0
        for i in range(256):
            # 确保K的索引有效，i % keylen 不会超出K的范围
            j = (j + self.S[i] + K[i % keylen]) % 256
            self.S[i], self.S[j] = self.S[j], self.S[i]
    
    def transform(self, data: bytes) -> bytes:
        """ 加密或解密数据 """
        i = 0
        j = 0
        output = bytearray()
        S = self.S.copy()  # 使用副本来保持原始S盒状态不变
        
        for k in range(len(data)):
            i = (i + 1) % 256
            j = (j + S[i]) % 256
            S[i], S[j] = S[j], S[i]
            key = S[(S[i] + S[j]) % 256]
            output.append(key ^ data[k])
        
        return bytes(output)

if __name__ == "__main__":
    data = b"wednesday is shit"
    print("明文:", data.decode())
    
    # 注意：C++版本密钥包含结尾的null，这里显式添加\x00
    K = 'key'.encode()
    keylen = len(K)  # 12字节（包含null）
    
    rc4_encrypt = RC4()
    rc4_decrypt = RC4()
    
    rc4_encrypt.set_key(K, keylen)
    rc4_decrypt.set_key(K, keylen)
    
    # 加密
    encrypted = rc4_encrypt.transform(data)

    print("密文:", encrypted.hex())  # 使用hex显示二进制更清晰
    
    # 解密
    decrypted = rc4_decrypt.transform(encrypted)
    print("解密后明文:", decrypted.decode())